extension.ts

```diff
@@ -0,0 +1,374 @@
+    id: `vscode-search-provider-${uri}`,
+    // TODO: Check the outcome of get_basename on file URIs
+    name: file.get_basename() || `<unnamed ${kind}>`,
+    // TODO: Test what this parse name thing actually is
+    readablePath: file.get_parse_name(),
+    kind,
+    file
+  };
+};
+
+/**
+ * Find recent items from VSCode.
+ *
+ * @param configDirectoryName The name of the config directory of the code app
+ * @returns A promise with recent items
+ */
+const findVSCodeRecentItems = (
+  configDirectoryName: string
+): Promise<RecentItems> =>
+  new Promise(resolve => {
+    // TODO: Use async load contents for async IO, to avoid blocking the shell
+    const contents = Gio.File.new_for_path(GLib.get_user_config_dir())
+      .get_child(configDirectoryName)
+      .get_child("storage.json")
+      .load_contents(null)[1];
+
+    const storage = JSON.parse(ByteArray.toString(contents));
+
+    const recentWorkspaceURIs: string[] =
+      storage.openedPathsList.workspaces3 ||
+      storage.openedPathsList.workspaces2 ||
+      storage.openedPathsList.workspaces ||
+      [];
+    const recentFileURIs: string[] =
+      storage.openedPathsList.files2 || storage.openedPathsList.files || [];
+
+    const recentItems = recentWorkspaceURIs
+      .map(createRecentItem("workspace"))
+      .concat(recentFileURIs.map(createRecentItem("file")));
+
+    resolve(new Map(recentItems.map(item => [item.id, item])));
+  });
+
+/**
+ * The registered provider
+ */
+// eslint-disable-next-line immutable/no-let
+let registeredProvider: "unregistered" | "registering" | SearchProvider =
+  "unregistered";
+
+/**
+ * Initialize this extension.
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/no-empty-function
+function init(): void {}
+
+/**
+ * Enable this extension.
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function enable(): void {
+  if (registeredProvider === "unregistered") {
+    const vscode = findVSCode();
+    if (vscode) {
+      registeredProvider = "registering";
+      findVSCodeRecentItems(vscode.configDirectoryName).then(
+        items => {
+          // If the user hasn't disabled the extension meanwhile create the
+          // search provider and registered it, both in our global variable
+          // and for gnome shell.
+          if (registeredProvider === "registering") {
+            registeredProvider = createProvider(
+              vscode.app,
+              ExtensionUtils.getSettings(),
+              items
+            );
+            Main.overview.viewSelector._searchResults._registerProvider(
+              registeredProvider
+            );
+          }
+        },
+        error => {
+          Main.notifyError(
+            "Failed to get recent VSCode entries",
+            error.message
+          );
+        }
+      );
+    } else {
+      Main.notifyError(
+        "VSCode not found",
+        "Try installing VSCode or VSCode OSS."
+      );
+    }
+  }
```

Comment:
I personally think this would be more readable using early exits.

Reply:
You're right, I'm not happy w/ this piece as well.  I'll refactor it :)

Reply:
Refactored, should look better now :)